services/src/main/java/org/fao/geonet/api/records/MetadataSharingApi.java

```diff
@@ -882,4 +942,206 @@ private void updateOwnership(Integer groupIdentifier,
+                                           boolean allowPublishInvalidMd, boolean allowPublishNonApprovedMd) throws Exception {
+        MetadataValidationRepository metadataValidationRepository = context.getBean(MetadataValidationRepository.class);
+        IMetadataValidator validator = context.getBean(IMetadataValidator.class);
+        IMetadataStatus metadataStatusRepository = context.getBean(IMetadataStatus.class);
+
+        if (!allowPublishInvalidMd) {
+            boolean hasValidation =
+                (metadataValidationRepository.count(MetadataValidationSpecs.hasMetadataId(metadata.getId())) > 0);
+
+            if (!hasValidation) {
+                validator.doValidate(metadata, context.getLanguage());
+                dm.indexMetadata(metadata.getId() + "", true, null);
+            }
+
+            boolean isInvalid =
+                (metadataValidationRepository.count(MetadataValidationSpecs.isInvalidAndRequiredForMetadata(metadata.getId())) > 0);
+
+            if (isInvalid) {
+                throw new Exception("The metadata " + metadata.getUuid() + " it's not valid, can't be published.");
+            }
+        }
+
+        if (!allowPublishNonApprovedMd) {
+            MetadataStatus metadataStatus = metadataStatusRepository.getStatus(metadata.getId());
+
+            String statusId = metadataStatus.getId().getStatusId() + "";
+            boolean isApproved = statusId.equals(StatusValue.Status.APPROVED);
+
+            if (!isApproved) {
+                throw new Exception("The metadata " + metadata.getUuid() + " it's not approved, can't be published.");
+            }
+        }
+
+    }
+
+
+    /**
+     * Shares a metadata with the {@link ReservedGroup#all} group to publish/unpublish it.
+     *
+     * @param metadataUuid  Metadata uuid.
+     * @param publish       Flag to publish/unpublish the metadata.
+     * @param request
+     * @param session
+     * @throws Exception
+     */
+    private void shareMetadataWithAllGroup(String metadataUuid, boolean publish,
+                                   HttpSession session, HttpServletRequest request) throws Exception {
+        AbstractMetadata metadata = ApiUtils.canEditRecord(metadataUuid, request);
+        ApplicationContext appContext = ApplicationContextHolder.get();
+        ServiceContext context = ApiUtils.createServiceContext(request);
+
+
+        //--- in case of owner, privileges for groups 0,1 and GUEST are disabled
+        //--- and are not sent to the server. So we cannot remove them
+        UserSession us = ApiUtils.getUserSession(session);
+        boolean isAdmin = Profile.Administrator == us.getProfile();
+        boolean isReviewer = Profile.Reviewer == us.getProfile();
+        if (us.getUserIdAsInt() == metadata.getSourceInfo().getOwner() &&
+            !isAdmin &&
+            !isReviewer) {
+            throw new Exception("User not allowed to publish the metadata " + metadataUuid);
```

Comment:
This is wrong @PascalLike. If user is a reviewer of the group of the record, then it is allowed. Right ?

Reply:
Maybe to replace by `AccessManage#hasReviewPermission` instead ?